python/tempo/tsdf.py

```diff
@@ -36,121 +65,217 @@ class TSDF:
     def __init__(
         self,
         df: DataFrame,
-        ts_col: str = "event_ts",
-        partition_cols: Optional[list[str]] = None,
-        sequence_col: Optional[str] = None,
-    ):
+        ts_schema: TSSchema = None,
+        ts_col: str = None,
+        series_ids: Collection[str] = None,
+        validate_schema=True,
+    ) -> None:
+        self.df = df
+        # construct schema if we don't already have one
+        if ts_schema:
+            self.ts_schema = ts_schema
+        else:
+            self.ts_schema = TSSchema.fromDFSchema(self.df.schema, ts_col, series_ids)
+        # validate that this schema works for this DataFrame
+        if validate_schema:
+            self.ts_schema.validate(df.schema)
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def __str__(self) -> str:
+        return f"""TSDF({id(self)}):
+    TS Index: {self.ts_index}
+    Series IDs: {self.series_ids}
+    Observational Cols: {self.observational_cols}
+    DataFrame: {self.df.schema}"""
+
+    def __withTransformedDF(self, new_df: DataFrame) -> "TSDF":
         """
-        Constructor
-        :param df:
-        :param ts_col:
-        :param partition_cols:
-        :sequence_col every tsdf allows for a tie-breaker secondary sort key
-        """
-        self.ts_col = self.__validated_column(df, ts_col)
-        self.partitionCols = (
-            []
-            if partition_cols is None
-            else self.__validated_columns(df, partition_cols.copy())
+        This helper function will create a new :class:`TSDF` using the current schema, but a new / transformed :class:`DataFrame`
+
+        :param new_df: the new / transformed :class:`DataFrame` to
+
+        :return: a new TSDF object with the transformed DataFrame
+        """
+        return TSDF(new_df, ts_schema=deepcopy(self.ts_schema), validate_schema=False)
+
+    def __withStandardizedColOrder(self) -> TSDF:
+        """
+        Standardizes the column ordering as such:
+        * series_ids,
+        * ts_index,
+        * observation columns
+
+        :return: a :class:`TSDF` with the columns reordered into "standard order" (as described above)
+        """
+        std_ordered_cols = (
+            list(self.series_ids) + [self.ts_index.name] + list(self.observational_cols)
         )
 
-        self.df = df
-        self.sequence_col = "" if sequence_col is None else sequence_col
+        return self.__withTransformedDF(self.df.select(std_ordered_cols))
 
-        # Add customized check for string type for the timestamp.
-        # If we see a string, we will proactively created a double
-        # version of the string timestamp for sorting purposes and
-        # rename to ts_col
+    @classmethod
```

Comment:
I guess I'm struggling to understand why we'd instantiate an instance of TSDF to do a df -> df operation. Seems like this could be a static method other than it's protected. And if it's protected, do we want to expose it as a class method?